src/java/org/apache/cassandra/io/sstable/format/bti/BtiTableReader.java

```diff
@@ -0,0 +1,558 @@
+import static org.apache.cassandra.utils.concurrent.SharedCloseable.sharedCopyOrNull;
+
+/**
+ * Reader of SSTable files in BTI format (see {@link BtiFormat}), written by {@link BtiTableWriter}.
+ *
+ * SSTableReaders are open()ed by Keyspace.onStart; after that they are created by SSTableWriter.renameAndOpen.
+ * Do not re-call open() on existing SSTable files; use the references kept by ColumnFamilyStore post-start instead.
+ */
+public class BtiTableReader extends SSTableReaderWithFilter
+{
+    private final FileHandle rowIndexFile;
+    private final PartitionIndex partitionIndex;
+
+    public BtiTableReader(Builder builder, SSTable.Owner owner)
+    {
+        super(builder, owner);
+        this.rowIndexFile = builder.getRowIndexFile();
+        this.partitionIndex = builder.getPartitionIndex();
+    }
+
+    protected final Builder unbuildTo(Builder builder, boolean sharedCopy)
+    {
+        Builder b = super.unbuildTo(builder, sharedCopy);
+        if (builder.getPartitionIndex() == null)
+            b.setPartitionIndex(sharedCopy ? sharedCopyOrNull(partitionIndex) : partitionIndex);
+        if (builder.getRowIndexFile() == null)
+            b.setRowIndexFile(sharedCopy ? sharedCopyOrNull(rowIndexFile) : rowIndexFile);
+
+        return b;
+    }
+
+    @Override
+    protected List<AutoCloseable> setupInstance(boolean trackHotness)
+    {
+        ArrayList<AutoCloseable> closeables = Lists.newArrayList(rowIndexFile, partitionIndex);
+        closeables.addAll(super.setupInstance(trackHotness));
+        return closeables;
+    }
+
+    /**
+     * Whether to filter out data before {@code first}. Needed for sources of data in a compaction, where the relevant
+     * output is opened early -- in this case the sstable's start is changed, but the data can still be found in the
+     * file. Range and point queries must filter it out.
+     */
+    protected boolean filterFirst()
+    {
+        return openReason == OpenReason.MOVED_START;
+    }
+
+    /**
+     * Whether to filter out data after {@code last}. Early-open sstables may contain data beyond the switch point
+     * (because an early-opened sstable is not ready until buffers have been flushed), and leaving that data visible
+     * will give a redundant copy with all associated overheads.
+     */
+    protected boolean filterLast()
+    {
+        return openReason == OpenReason.EARLY && partitionIndex instanceof PartitionIndexEarly;
+    }
+
+    public long estimatedKeys()
+    {
+        return partitionIndex == null ? 0 : partitionIndex.size();
+    }
+
+    @Override
+    protected TrieIndexEntry getRowIndexEntry(PartitionPosition key,
+                                              Operator operator,
+                                              boolean updateStats,
+                                              boolean permitMatchPastLast,
+                                              SSTableReadsListener listener)
+    {
+        assert !permitMatchPastLast;
```

Comment:
nit: Do we need this, given we never use it anyway? ...or does it provide some sort of protection against aberrant usage?